### ISSUE TITLE
Contingency Table API

### ISSUE THREAD

**Issue by ebolyen** [biocore/scikit-bio#848]
This will be a place to discuss the specific features that a contingency table should be able to support and what that API will looks like.

**Comment by mortonjt** [comment]
This may be a stupid question about the ```skbio.io.biom``` registry mentioned before.

But isn't BIOM backwards compatible?  It seems to me that it would be more straightforward to use the BIOM parsers and create the biom.Table object.

**Comment by mortonjt** [comment]
Also I remember there was some discussion about whether to use Pandas or the biom.Table object as the core data type.  Here are the pros and cons that I see

__pandas.DataFrame__
Pros
* Dense matrix storage (could be easier for users to manipulate)
* Nice text visualization
* Has apply() function

Cons
* No collapse() functionality ( this could be added - but it could get messy )
* No support for 2D metadata (can be remedied with 3 pandas objects)

__biom.Table__
Pros
* Sparse matrix storage.  Lenient on memory. Could give a speed boost via sparse matrix manipulations
* Nice support for 2D metadata

Cons
* No apply() function (would need to build it from scratch)
* Visualization of raw data can overflow screen buffer

I'm still in favor of having the BIOM table as the underlying datatype.  I don't think it would be too difficult to add in an apply() function.  And it already has many similar functions supported by Pandas.  But I don't have particularly strong opinions - I think both solutions are reasonable.

If we do want to go the Pandas route - converting from a biom.Table to pandas.DataFrame is fairly straightforward

Its just
```
from biom import Table
import pandas as pd
table = load_table("some_table.biom")
mat = table._get_sparse_data().toarray().transpose()
df = pd.DataFrame(mat)
df.index = table.ids(axis='sample')
df.columns = table.ids(axis='observation')
```

**Comment by Jorge-C** [comment]
@mortonjt pandas has a [sparse backend too](http://pandas.pydata.org/pandas-docs/dev/sparse.html).

**Comment by mortonjt** [comment]
Hmm.  I have never used that feature.  Good catch @Jorge-C !

**Comment by gregcaporaso** [comment]
@mortonjt, note that Pandas does have [GroupBy](http://pandas.pydata.org/pandas-docs/dev/groupby.html), which I think gets the ``collapse`` functionality that you mention.

Can you give an example of what you mean by 2D metadata?

**Comment by mortonjt** [comment]
Pardon my sloppy notation. When I meant by 2D metadata. I was just referring to observation and sample metadata found in biom (aka zip codes and taxonomy information)

Yes, I am familiar with the pandas group by operation. I actually used it to perform the collapse function before I became familiar with biom. 

My major concern with the pandas approach is figuring out how to manage the metadata.  I think merging all of the metadata and abundance data into a single dataframe will cause much heartache later.

**Comment by gregcaporaso** [comment]
Ok, got it. It's definitely not immediately obvious what the right option is. I think mutliple dataframes is one option (as you point out). Using the multi-index might work out too, but I do think multiple dataframes is probably better. 

Re: groupby/collapse, if we went the DataFrame-under-the-hood route what I think we'd want to do is write our own ``collapse`` that wraps ``GroupBy`` to simplify the functionality that we use regularly. That would then simplify the interface for what we need to do regularly, but also give us the power to make use of the more advanced pandas functionality.

**Comment by mortonjt** [comment]
Having a multi-index is not a bad idea.  However, doing so would enforce the following design constraints

* DataFrames cannot have other metadata information for OTUs (forget about adding gram positive / gram negative labels for OTUs)
* DataFrames can only have a specific biological data type (cannot merge OTUs with KEGGs).  This could be a major roadblock for statistical tools. I can think of scenarios where I would want to simultaneously analyze different biological datasets and having them in a single contingency table would be useful. 

I know that the collapse function is very doable using pandas.  So that shouldn't be an issue.

**Comment by ebolyen** [comment]
I appreciate all the interest in this topic, but if we could keep the discussion focused on what API features would be useful, and less about what certain implementation decisions make convenient/difficult, that would be great! The topic seems to be trending in that direction anyways, so thank you all.

To directly address why we wouldn't just import the biom package: 
There is no need (for us) to maintain a core bioinformatics object external to scikit-bio. It would be easier to maintain API compatibility and ensure our deprecation process is viable if the concept of biom.table was migrated instead of wrapped. This also allows the biom-format to focus more on what needs to exist in the file format and to create some useful command line tools with scikit-bio without needing to worry about API management of the biom.table object (these considerations would be routine for scikit-bio anyways). Basically, we won't be duplicating code so much as deprecating biom.table in favor of scikit-bio's contingency table (there is likely to be a fair amount of logic and API migrated in this endeavour) 

But before we do any of that, we need to figure out what a contingency table can do, what those operations look like, and why those are the best operations to provide. Let's keep that the focus of this discussion from here on out. Feel free to open an issue to discuss implementation, but I feel that is premature at this moment (next week would be a better time to worry about it). Discussions about  the pandas and biom.table API are of course completely encouraged as they can serve as a model for our contingency table.

**Comment by mortonjt** [comment]
That makes sense.  

I do agree with @ebolyen .  It would be wise not to rush through this.

At this point, I'm willing to buy that pandas may be the way to go.

As far as API functions go, here are a few functions that I think would be key to implement.  I'm going to try to rank them by importance

* ```read()```
* ```write()```
* ```apply( func, axis )```
* ```filter(func, axis)```
* ```groupby( id, axis )```
* ```merge(table)```
* ```collapse( func, axis )```
* ```sort( axis, index [optional] )```
* ```__str__()```

Not sure about how indexing data/metadata should be handled.  Any thoughts?

**Comment by mortonjt** [comment]
@ebolyen , I thought a little more about what you said.

We could be potentially shooting ourselves in foot if we move ahead and try to rewrite all of the biom parsers for each biom version.  For *every* future release of biom, we would have to write up a new set of parsers.  To me, this seems to be waste of developer time.

Instead, I would think it would be more manageable to import BIOM, but have a single conversion method that would convert the BIOM object to our new contingency table object.  That way, we could take full advantage of BIOM backwards compatibility while keeping the APIs relatively separate

**Comment by ebolyen** [comment]
@mortonjt BIOM is honestly less backwards compatible than you might think. Additionally it is largely the same development team either way, so those parses have to be written one way or another if the format is updated again. What scikit-bio does provide is a very robust IO mechanism which should actually make identifying and parsing different variations of BIOM much simpler than it is right now.

**Comment by josenavas** [comment]
:+1: to @ebolyen

**Comment by mortonjt** [comment]
That makes sense.  Thanks for clarifying @ebolyen !

**Comment by mortonjt** [comment]
At this point, it seems that we agree on the following

* The Contingency table should be completely separate from the BIOM API
* BIOM parsers will need to be built
* A ```pandas.DataFrame``` should be used as the core data type

I thought a bit about @gregcaporaso comment about pandas multiIndex and I actually think it is a really good idea - it naturally allows for an intuitive display/organization of taxonomies.

However, if a MultiIndex is to be used, the following should be considered
* Do we care about other metadata for OTUs other than taxonomies?  For example - would we want to store information about gram positive vs. gram negative for each OTU
* Do we care about storing different data types (e.g. store metabolites, genes and OTUs all in the same table)?

I don't think these issues are insurmountable when designing the table - but they are things to consider during implementation.

The other option is to use multiple DataFrames to store the metadata, or a combination of the two approaches.

**Comment by ebolyen** [comment]
Thanks for bringing up the interesting point about observations being of different types, @mortonjt. Also, keep in mind that we aren't really worried about how to do something, it may be true that a pandas dataframe is convenient, but we may also find a numpy array is all we need, either way, that is a decision to make after we decide what we want to accomplish, rather than constraining ourselves off the bat to what a particular implementation makes easy or hard. 

Back to what you brought up regarding different data types:

I think the ability to have polymorphic observations within a table may throw off some of the statistics, which assume observations are equally weighted. That being said, I don't think ContingencyTable should ever particularly care what an observation is. It's primary focus is the contingency itself between observations and samples, not necessarily what either of them represent. 

Another concept we are starting to toy with is the idea that per-sample metadata doesn't always belong to an object, as much as an ID. In an effort to reduce duplication of ID validation, it occurred to us that a discrete ID and IDCollection/Map/Whatever class may be useful. We don't know if this is a user facing abstraction yet, but it would seem that metadata about an observation and a sample could fit well with this. Something I think might be even more powerful, is if the ID represented an enforceable relationship. So instead of storing taxonomy as metadata on an observation, you would instead associate a taxonomy with an ID. ID's may even be able to track collapsing and renaming (like sequence demultiplexing).

**Comment by mortonjt** [comment]
I like that idea about associating taxonomy with an id
Here are a couple of questions

* Would we want to enforce symmetry between sample metadata and observation metadata?  For example, it makes sense to collapse observations based on taxonomy.  But would we interested in collapsing metadata based on location for instance (aka collapsing samples by state or country).
* Would we want to have different types of metadata? Since there are metadata fields that cannot be associated with an ID, would it make sense to separate these fields into two groups - fields associated with ID and other descriptor fields?

**Comment by wasade** [comment]
Yes, in the case of sample replicates or even sample taxonomy (in the case
of comparative genomics)

**Comment by mortonjt** [comment]
Ok.  That would complicate things a bit.  That would implicate the multiple metadata fields could serve as an index for observations/samples.  In addition, how would samples be combined?  That would imply that there is some normalization scheme allowing samples to be compared to each other.  As far as I can tell, that is an unanswered question.

Let me take a step back and pose the following question.
* What key features are missing in the BIOM API that we think are crucial in the Contingency table?

I imagine there must be several insurmountable key features missing - otherwise we could have just merged/ported the entire BIOM API into skbio

**Comment by Jorge-C** [comment]
:+1: being able to answer that question will get us much closer to a proposal that is both concrete and improves on the state of the art.

**Comment by wasade** [comment]
@gregcaporaso, what repo did the code for handling sample replicates go into? Can't find it right now

@mortonjt, looking through, potentially just `apply`, although it is nearly the same as `transform` with the exception that you might want to apply to the whole table, not just per sample or observation. 

To be honest, we have spent a _lot_ of time over the last few years refining the API, in particular, this past spring. The biggest thing I can see is perhaps changing how we manage metadata, particularly group metadata (e.g., a phylogenetic tree). But, it isn't clear to me if that necessitates API changes and I don't think that should be a blocker for getting the existing API in place with maybe some method name changes, and of course, pulling out IO operations.

**Comment by Jorge-C** [comment]
I agree with @wasade's comments. What are the drawbacks of porting biom's table with its current API / reasons to start from scratch?

**Comment by gregcaporaso** [comment]
I agree that we should discuss this. Could this be a discussion topic for
next week's sprint?

**Comment by antgonza** [comment]
Yup, we should put all issues we want to discuss in the hackpad

**Comment by gregcaporaso** [comment]
Thanks for the input everyone, we've been discussing this and are thinking that we should hold off on this for now as there are other higher priority issues. We can discuss further next week if we want more discussion at this time.
